File: mgi/src/controllers/authController.ts
import { Request, Response, NextFunction, CookieOptions } from 'express';
import { Inject, Service } from 'typedi';
import config from '../../config';

import IAuthController from './IControllers/IAuthController';
import IAuthService from '../services/IServices/IAuthService';
import { IUserWithRoleDTO } from '../dto/IUserDTO';
import { Result } from '../core/logic/Result';

@Service()
export default class AuthController implements IAuthController /* TODO: extends ../core/infra/BaseController */ {
    constructor(@Inject(config.services.auth.name) private authServiceInstance: IAuthService) {}

    public async login(req: Request, res: Response, next: NextFunction) {
        try {
            const result = await this.authServiceInstance.login(req.body.email, req.body.password);

            if (result.isFailure) {
                return res.status(400).send({ error: result.errorValue() });
            }

            const { token, userDTO } = result.getValue();

            res.header('Access-Control-Allow-Credentials', 'true');
            res.header('Access-Control-Allow-Origin', config.clientURL);


            this.setCookie(res, token);

            return res.json(userDTO).status(201);
        } catch (e) {
            return next(e);
        }
    }

    public async signup(req: Request, res: Response, next: NextFunction) {
        try {
            const userOrError = (await this.authServiceInstance.signup(req.body)) as Result<{
                userDTO: IUserWithRoleDTO;
                token: string;
            }>;

            if (userOrError.isFailure) {
                return res.status(400).send({ error: userOrError.errorValue() });
            }

            const { token, userDTO } = userOrError.getValue();

            this.setCookie(res, token);

            return res.json(userDTO).status(201);
        } catch (e) {
            return next(e);
        }
    }

    public async logout(req: Request, res: Response, next: NextFunction) {
        try {
            res.header('Access-Control-Allow-Credentials', 'true');
            res.header('Access-Control-Allow-Origin', config.clientURL);
            
            // destroy cookie
            res.clearCookie(config.cookieName);
            return res.json({ message: 'logged out successfully!' }).status(201);
        } catch (e) {
            return next(e);
        }
    }

    public async session(req: Request, res: Response, next: NextFunction) {
        try {
            res.header('Access-Control-Allow-Credentials', 'true');
            res.header('Access-Control-Allow-Origin', config.clientURL);

            const token = req.cookies[config.cookieName];            

            if (!token) {
                return res.status(401).json({ message: 'Unauthorized: Missing token' });
            }

            const result = await this.authServiceInstance.session(token);

            if (result.isFailure) {
                return res.status(400).send({ error: result.errorValue() });
            }

            const userDTO = result.getValue();

            return res.json(userDTO).status(201);
        } catch (e) {
            return next(e);
        }
    }

    private setCookie(res: Response, token: string) {
        let cookieOptions: CookieOptions = {};

        if (process.env.NODE_ENV === 'production') {
            cookieOptions = {
                httpOnly: true,
                sameSite: 'none', // Allows cross-origin cookies
                secure: true, // Requires HTTPS if in production
                maxAge: 1000 * 60 * 60 * 24 * 7, // 7 days
                path: '/',
            };
        } else {
            cookieOptions = {
                httpOnly: true,
                path: '/', // path = where the cookie is valid
                domain: 'localhost', // domain = what domain the cookie is valid on
                secure: false,
                sameSite: "lax", // "strict" | "lax" | "none" (secure must be true)
                maxAge: 1000 * 60 * 60 * 24 * 7, // 7 days
            };
        }

        res.cookie(config.cookieName, token, cookieOptions);
    }
}

File: mpt/Controllers/AuthController.cs
using Microsoft.AspNetCore.Mvc;
using Mpt.IServices;
using Mpt.Dtos;
using Mpt.IControllers;
using Microsoft.AspNetCore.Authorization;
using Newtonsoft.Json;


namespace Mpt.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class AuthController : ControllerBase, IAuthController
    {
        private readonly IConfiguration _config;
        private readonly IAuthService _service;
        private readonly IWebHostEnvironment _env;

        public AuthController(IConfiguration config, IAuthService service, IWebHostEnvironment env)
        {
            this._config = config;
            this._service = service;
            this._env = env;
        }

        // POST: api/Auth/login
        [AllowAnonymous]
        [HttpPost("login")]
        public async Task<ActionResult<UserWithRoleDto>> Login(UserLoginDto loginDto)
        {
            try
            {
                var user = await _service.LoginAsync(loginDto);

                if (user.IsFailure)
                {
                    return BadRequest(new { error = user.Error });
                }

                var token = this._service.GenerateJwtToken(user.Value);
                if (token.IsFailure)
                {
                    return BadRequest(new { error = token.Error });
                }

                SetCookie(token.GetValue());
                return Ok(user.GetValue());
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return BadRequest(new { error = ex.Message });
            }
        }

        // POST: api/Auth
        [AllowAnonymous]
        [HttpPost("signup")]
        public async Task<ActionResult<UserWithRoleDto>> Signup(SignupUserDto dto)
        {
            try
            {
                var user = await _service.SignupAsync(dto);

                if (user.IsFailure)
                {
                    return BadRequest(new { error = user.Error });
                }

                // var token = this._service.GenerateJwtToken(user.Value);
                // if (token.IsFailure)
                // {
                //     return BadRequest(new { error = token.Error });
                // }

                // SetCookie(token.GetValue());

                return Ok(user.GetValue());

            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return BadRequest(new { error = ex.Message });
            }
        }

        // GET: api/Auth/session
        [AllowAnonymous]
        [HttpGet("session")]
        public async Task<ActionResult<UserWithRoleDto>> Session()
        {
            try
            {
                var cName = GetCookieName();
                var token = Request.Cookies[cName];

                if (string.IsNullOrEmpty(token))
                {
                    return BadRequest(new { message = "No session found" });
                }

                var user = await _service.SessionAsync(token);

                if (user.IsFailure)
                {
                    return BadRequest(new { error = user.Error });
                }

                return Ok(user.GetValue());
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return BadRequest(new { error = ex.Message });
            }
        }

        // POST: api/Auth/validate
        [AllowAnonymous]
        [HttpPost("validate")]
        public async Task<ActionResult<UserWithRoleDto>> ValidateToken(TokenDto token)
        {
            try
            {
                var user = await _service.SessionAsync(token.Token);

                if (user.IsFailure)
                {
                    return BadRequest(new { error = user.Error });
                }

                return Ok(user.GetValue());
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return BadRequest(new { error = ex.Message });
            }
        }

        // POST: apit/Auth/logout
        [AllowAnonymous]
        [HttpPost("logout")]
        public ActionResult Logout()
        {
            try
            {
                // destroy cookie
                var cName = GetCookieName();
                Response.Cookies.Delete(cName);

                return Ok(new { message = "Logout successfully" });
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return BadRequest(new { error = ex.Message });
            }
        }

        private void SetCookie(string token)
        {
            var cName = GetCookieName();

            var cookieOptions = new CookieOptions
            {
                HttpOnly = true,
                Expires = DateTime.UtcNow.AddDays(7),
                SameSite = SameSiteMode.None,
                Secure = true,
                Path = "/",
            };

            if (this._env.IsDevelopment())
            {
                cookieOptions.Secure = false;
                cookieOptions.SameSite = SameSiteMode.Lax;
                cookieOptions.Domain = "localhost";
            }

            Response.Cookies.Append(cName, token, cookieOptions);
        }

        private string GetCookieName()
        {
            return this._config.GetValue<string>("Cookie:name") ?? "robdronego_authCookie";
        }

    }
}